eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala

```diff
@@ -0,0 +1,293 @@
+                                     pluginTimeout: FiniteDuration,
+                                     nodeParams: NodeParams,
+                                     wallet: OnchainPubkeyCache,
+                                     context: ActorContext[OpenChannelInterceptor.Command]) {
+
+  import OpenChannelInterceptor._
+
+  private def waitForRequest(): Behavior[Command] = {
+    receiveCommandMessage[WaitForRequestCommands](context, "waitForRequest") {
+      case initiator: OpenChannelInitiator => sanityCheckInitiator(initiator)
+      case nonInitiator: OpenChannelNonInitiator => checkRateLimits(nonInitiator)
+    }
+  }
+
+  private def sanityCheckInitiator(initiator: OpenChannelInitiator): Behavior[Command] = {
+    if (initiator.open.fundingAmount >= Channel.MAX_FUNDING && !initiator.localFeatures.hasFeature(Wumbo)) {
+      initiator.replyTo ! Status.Failure(new RuntimeException(s"fundingAmount=${initiator.open.fundingAmount} is too big, you must enable large channels support in 'eclair.features' to use funding above ${Channel.MAX_FUNDING} (see eclair.conf)"))
+      waitForRequest()
+    } else if (initiator.open.fundingAmount >= Channel.MAX_FUNDING && !initiator.remoteFeatures.hasFeature(Wumbo)) {
+      initiator.replyTo ! Status.Failure(new RuntimeException(s"fundingAmount=${initiator.open.fundingAmount} is too big, the remote peer doesn't support wumbo"))
+      waitForRequest()
+    } else if (initiator.open.fundingAmount > nodeParams.channelConf.maxFundingSatoshis) {
+      initiator.replyTo ! Status.Failure(new RuntimeException(s"fundingAmount=${initiator.open.fundingAmount} is too big for the current settings, increase 'eclair.max-funding-satoshis' (see eclair.conf)"))
+      waitForRequest()
+    } else {
+      // If a channel type was provided, we directly use it instead of computing it based on local and remote features.
+      val channelFlags = initiator.open.channelFlags_opt.getOrElse(nodeParams.channelConf.channelFlags)
+      val channelType = initiator.open.channelType_opt.getOrElse(ChannelTypes.defaultFromFeatures(initiator.localFeatures, initiator.remoteFeatures, channelFlags.announceChannel))
+      val dualFunded = Features.canUseFeature(initiator.localFeatures, initiator.remoteFeatures, Features.DualFunding)
+      val upfrontShutdownScript = Features.canUseFeature(initiator.localFeatures, initiator.remoteFeatures, Features.UpfrontShutdownScript)
+      val localParams = createLocalParams(nodeParams, initiator.localFeatures, upfrontShutdownScript, channelType, isInitiator = true, dualFunded = dualFunded, initiator.open.fundingAmount, initiator.open.disableMaxHtlcValueInFlight)
+      peer ! Peer.SpawnChannelInitiator(initiator.open, ChannelConfig.standard, channelType, localParams, initiator.replyTo.toClassic)
+      waitForRequest()
+    }
+  }
+
+  private def checkRateLimits(nonInitiator: OpenChannelNonInitiator): Behavior[Command] = {
```

Comment:
This should be the last thing we do, otherwise if the `PendingChannelsRateLimiter` accepts the channel but then we reject it for another reason, that channel will still use a slot in the `PendingChannelsRateLimiter` state that won't be freed, and at some point we'll end up rejecting all requests only because of such "zombie channels".

Reply:
Now that `PendingChannelRateLimiter` responds to `ChannelAborted` I think it makes sense to have `OpenChannelInterceptor` emit `ChannelAborted` in any situation where the channel is not accepted. This will fix the problem with "zombie channels" while still performing the rate limit check before the plugin.

To my mind it makes sense to do rate limiting upstream of running a plugin so that the plugin can depend on any channel being opened that it accepts.

See [024d78a](https://github.com/ACINQ/eclair/pull/2552/commits/024d78a8f7dc6b3cd8d57f7ed4cd8c486b2bcd67).

What do you think?

Reply:
Note: should I add a check to `PendingChannelRateLimiter` that rejects any attempt to open a channel with a `temporaryChannelId` that is already being tracked?

Otherwise someone could request a channel and then request another one with the same `temporaryChannelId` without it counting towards the rate limit. They could also fail the second request and remove both entries from `PendingChannelRateLimiter`. Generally, it's easier to reason about if we know there can be no duplicate channel ids.

Reply:
Good idea, that makes sense!

Reply:
> Note: should I add a check to PendingChannelRateLimiter that rejects any attempt to open a channel with a temporaryChannelId that is already being tracked?

No, that should be rejected even before that, and it should already be the case, the `Peer` is supposed to check that `temporaryChannelId` isn't duplicated.

Reply:
> > Note: should I add a check to PendingChannelRateLimiter that rejects any attempt to open a channel with a temporaryChannelId that is already being tracked?
> 
> No, that should be rejected even before that, and it should already be the case, the `Peer` is supposed to check that `temporaryChannelId` isn't duplicated.

Ha! Of course, that's pretty much all `Peer` does when it gets a new channel open request. I was over thinking things.